Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp

```diff
@@ -459,7 +458,7 @@ DOM::ExceptionOr<void> XMLHttpRequest::open(String const& method, String const&
     // FIXME: 2. If settingsObject has a responsible document and it is not fully active, then throw an "InvalidStateError" DOMException.
 
     // 3. If method is not a method, then throw a "SyntaxError" DOMException.
-    if (!is_method(method))
+    if (!is_token(method))
```

Comment:
This is now very confusing considering the comment above it. Why change that it nicely maps to a spec concept?

Reply:
I was very confused by this one myself. Going through the fetch spec and this section https://datatracker.ietf.org/doc/html/rfc7230#section-3.1.1 in the reffered RFC several times I interpreted the RFC such that both method and header names are actually what falls under the concept token. Therefore I renamed it to is_token() and use it to both validate header names and methods.

Reply:
afaict "method" is only grammatically equivalent, there's a valid list of methods, but no valid list of tokens (which makes sense)

>     method         = token
> The request methods defined by this specification can be found in
> [Section 4 of [RFC7231]](https://datatracker.ietf.org/doc/html/rfc7231#section-4)


while a token is just a series of characters with no particular semantics:
>     token          = 1*tchar
>     tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
>                     / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
>                     / DIGIT / ALPHA
>                     ; any VCHAR, except delimiters

so this function _should_ be called `is_method`.

Reply:
This was discussed in this PR as well: https://github.com/SerenityOS/serenity/pull/13476#discussion_r841597657.

You can have custom methods, the only requirement is that standardized methods are upper-cased which is done in `normalize_method()`.

Reply:
Not sure how that's related :)

All the spec says for "method" is:

> https://fetch.spec.whatwg.org/#concept-method
> A method is a [byte sequence](https://infra.spec.whatwg.org/#byte-sequence) that matches the [method](https://datatracker.ietf.org/doc/html/rfc7230#section-3.1.1) token production.

which links to

> https://datatracker.ietf.org/doc/html/rfc7230#section-3.1.1
>   The method token indicates the request method to be performed on the target resource.  The request method is case-sensitive.
>
> ```
>      method         = token
>      token          = 1*tchar
>      tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
>                     / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
>                     / DIGIT / ALPHA
>                     ; any VCHAR, except delimiters
> ```

So "method" is correct here.

Reply:
I guess we are discussing different things. I was arguing for consolidation of method and header name production, but not being very clear about it. Since the two concepts are using the exact same regular expression.

I have updated it now from is_token() -> is_method(). I have also added a new method is_header_name() in commit: [7440fba](https://github.com/SerenityOS/serenity/pull/13520/commits/7440fba8fa08f2f45de9d07354a1d2e661fb1273)